site/content/blog/2020-12-04-common-i18n-mistakes.md

```diff
@@ -0,0 +1,315 @@
+
+```jsx
+let username = '@' + rootPost.username;
+if (!username.endsWith('s')) {
+    username += '\'s';
+}
+
+return (
+    <FormattedMessage
+        id='post.commentedOn'
+        defaultMessage='Commented on {username} message:'
+        values={{username}}
+    />
+)
+```
+
+Here, we're using react-intl to add Billy's name into the text as a value, but we're making a mistake here by trying to make the username into a possessive adjective ourselves. By manually adding the "'s", we're applying English grammar regardless of the user's language setting, so this label won't make sense in other languages.
+
+Instead, the "'s" should be part of the translation string. That required changing our grammar rules slightly since we previously avoided adding "'s" to the end of names that ended in the letter S, but the new way is still correct by other popular grammar guidelines. Isn't English great?
+
+```jsx
+return (
+    <FormattedMessage
+        id='post.commentedOn'
+        defaultMessage="Commented on @{username}'s message:"
+        values={{username: rootPost.username}}
+    />
+);
+```
+
+## Mistake 3: Incorrect pluralization
+
+Another common mistake is to forget that pluralization exists when counting something, so you end up with text that looks like "1 posts deleted". You can attempt to avoid that by adding (s) to the end of a word, but that doesn't look very nice, and other languages may not have anything comparable that they can do to replicate that. Thankfully, our translation libraries can help us out here.
+
+Say you're adding a feature that sends out an email to a number of people, and you want it to report back on its progress in both the UI and in the server's logs. You might write something like this:
+
+```jsx
+return (
+    <FormattedMessage
+        id='email_sender.remaining'
+        defaultMessage='{remaining, number} email(s) remaining.'
+        values={{remaining}}
+    />
+);
+```
+
+```go
+/*
+    The translation file contains:
+    {
+        "id": "app.email_sender.remaining",
+        "translation": "{{.Remaining}} email(s) remaining"
+    }
+*/
+
+func logEmailsRemaining(remaining int) {
+    log.Print(translateFunc("app.email_sender.remaining", map[string]interface{}{
+        "Remaining": remaining,
+    }))
+}
+```
+
+Note that the React example uses a feature of react-intl to specify that `remaining` is a number. Without it, "0 emails remaining" wouldn't be shown properly because it's a falsey value in JavaScript.
+
+That is an example of how we can tell react-intl to behave slightly differently though. If we use the `plural` modifier, we can specify different text based on the value of `remaining`. go-i18n has a similar feature, but it takes two different versions of the translated string instead of constructing a one that is more complicated.
```

Comment:
```suggestion
That is an example of how we can tell react-intl to behave slightly differently though. If we use the `plural` modifier, we can specify different text based on the value of `remaining`. go-i18n has a similar feature, but it takes two different versions of the translated string instead of constructing one that's more complicated.
```